services/vision/mlvision/ml_model.go

```diff
@@ -0,0 +1,227 @@
+	resource.RegisterService(vision.Subtype, model, resource.Registration[vision.Service, *MLModelConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (vision.Service, error) {
+			attrs, err := resource.NativeConfig[*MLModelConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerMLModelVisionService(ctx, c.ResourceName().Name, attrs, actualR, logger)
+		},
+	})
+}
+
+// MLModelConfig specifies the parameters needed to turn an ML model into a vision Model.
+type MLModelConfig struct {
+	resource.TriviallyValidateConfig
+	ModelName string `json:"mlmodel_name"`
+}
+
+func registerMLModelVisionService(
+	ctx context.Context,
+	name string,
+	params *MLModelConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (vision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerMLModelVisionService")
+	defer span.End()
+
+	mlm, err := mlmodel.FromRobot(r, params.ModelName)
+	if err != nil {
+		return nil, err
+	}
+	classifierFunc, err := attemptToBuildClassifier(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+	}
+	detectorFunc, err := attemptToBuildDetector(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+	}
+	segmenter3DFunc, err := attemptToBuild3DSegmenter(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a 3D segmenter", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service 3D segmenter", params.ModelName)
+	}
+	// Don't return a close function, because you don't want to close the underlying ML service
+	return vision.NewService(name, r, nil, classifierFunc, detectorFunc, segmenter3DFunc)
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string, md mlmodel.MLMetadata) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch getTensorTypeFromName(name, md) {
+	case UInt8:
+		temp := me.([]uint8)
+		for _, t := range temp {
+			out = append(out, float64(t))
+		}
+	case Float32:
+		temp := me.([]float32)
+		for _, p := range temp {
+			out = append(out, float64(p))
+		}
+	default:
+		return nil
+	}
+	return out
+}
+
+// getTensorTypeFromName uses the metadata to find the expected type of the tensor.
+func getTensorTypeFromName(name string, md mlmodel.MLMetadata) string {
+	for _, o := range md.Outputs {
+		if strings.Contains(strings.ToLower(o.Name), strings.ToLower(name)) {
+			return o.DataType
+		}
+		if strings.Contains(name, "output") {
```

Comment:
this file has a lot of field names that are duplicated. you may want to put these in a const block to prevent future bugs around renames and it's generally good to single source once you are around 2-3 instances of a literal value

Reply:
Okay done.